akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashSpec.scala

```diff
@@ -22,98 +22,109 @@ object StashSpec {
+              Behaviors.same
+            case GetStashSize(replyTo) ⇒
+              replyTo ! buffer.size
+              Behaviors.same
+            case UnstashAll ⇒
+              buffer.unstashAll(ctx, active(processed))
+            case Unstash ⇒
+              log(ctx).debug(s"Unstash ${buffer.size}")
+              if (buffer.isEmpty)
+                active(processed)
+              else {
+                ctx.self ! Unstash // continue unstashing until buffer is empty
+                val numberOfMessages = 2
+                log(ctx).debug(s"Unstash $numberOfMessages of ${buffer.size}, starting with ${buffer.head}")
+                buffer.unstash(ctx, unstashing(processed), numberOfMessages, Unstashed)
+              }
+            case Stash ⇒
+              Behaviors.unhandled
+            case u: Unstashed ⇒
+              throw new IllegalStateException(s"Unexpected $u in stashing")
           }
-        case GetStashSize(replyTo) ⇒
-          replyTo ! buffer.size
-          Behaviors.same
-        case UnstashAll ⇒
-          Behaviors.unhandled
-        case u: Unstashed ⇒
-          throw new IllegalStateException(s"Unexpected $u in unstashing")
-      }
+        }
+
+      def unstashing(processed: Vector[String]): Behavior[Command] =
+        Behaviors.immutable { (ctx, cmd) ⇒
+          cmd match {
+            case Unstashed(msg: Msg) ⇒
+              log(ctx).debug(s"unstashed $msg")
+              unstashing(processed :+ msg.s)
+            case Unstashed(GetProcessed(replyTo)) ⇒
+              log(ctx).debug(s"unstashed GetProcessed")
+              replyTo ! processed
+              Behaviors.same
+            case msg: Msg ⇒
+              log(ctx).debug(s"got $msg in unstashing")
+              buffer.stash(msg)
+              Behaviors.same
+            case g: GetProcessed ⇒
+              log(ctx).debug(s"got GetProcessed in unstashing")
+              buffer.stash(g)
+              Behaviors.same
+            case Stash ⇒
+              stashing(processed)
+            case Unstash ⇒
+              if (buffer.isEmpty) {
+                log(ctx).debug(s"unstashing done")
+                active(processed)
+              } else {
+                ctx.self ! Unstash // continue unstashing until buffer is empty
+                val numberOfMessages = 2
+                log(ctx).debug(s"Unstash $numberOfMessages of ${buffer.size}, starting with ${buffer.head}")
+                buffer.unstash(ctx, unstashing(processed), numberOfMessages, Unstashed)
+              }
+            case GetStashSize(replyTo) ⇒
+              replyTo ! buffer.size
+              Behaviors.same
+            case UnstashAll ⇒
+              Behaviors.unhandled
+            case u: Unstashed ⇒
+              throw new IllegalStateException(s"Unexpected $u in unstashing")
+          }
+        }
+
+      active(Vector.empty)
     }
 
   class MutableStash(ctx: ActorContext[Command]) extends Behaviors.MutableBehavior[Command] {
```

Comment:
Not for this PR but i wonder if MutableBehavior should be in Behavior rather than Behaviors?

Reply:
can't because it's different for scaladsl and javadsl